besu/src/main/java/org/hyperledger/besu/services/PreMergeOnlineBlockDataPruner.java

```diff
@@ -0,0 +1,117 @@
+
+  private final int preMergeBlockPrunerRange;
+  private final long preMergeBlockPrunerSleepTime;
+  private final NetworkName networkName;
+  private final BlockchainStorage blockchainStorage;
+
+  /**
+   * Constructs a PreMergeOnlineBlockDataPruner with the supplied parameters
+   *
+   * @param preMergeBlockPrunerRange How many blocks to prune at a time
+   * @param preMergeBlockPrunerSleepTime How long to sleep between blocks in milliseconds
+   * @param networkName The network to target
+   * @param blockchainStorage The blockchain storage from which to prune blocks
+   */
+  public PreMergeOnlineBlockDataPruner(
+      final int preMergeBlockPrunerRange,
+      final long preMergeBlockPrunerSleepTime,
+      final NetworkName networkName,
+      final BlockchainStorage blockchainStorage) {
+    this.preMergeBlockPrunerRange = preMergeBlockPrunerRange;
+    this.preMergeBlockPrunerSleepTime = preMergeBlockPrunerSleepTime;
+    this.networkName = networkName;
+    this.blockchainStorage = blockchainStorage;
+  }
+
+  @Override
+  public void run() {
+    if (!SUPPORTED_NETWORKS.contains(networkName)) {
+      LOG.error(
+          "Unable to prune pre-merge blocks and transaction receipts for network: {}", networkName);
+      return;
+    }
+    final long mergeBlockNumber = getMergeBlockNumber(networkName);
+
+    LOG.info("Pruning pre-merge blocks and transaction receipts");
+    for (long startBlockNumber = 0;
+        startBlockNumber < mergeBlockNumber;
+        startBlockNumber += preMergeBlockPrunerRange) {
+      BlockchainStorage.Updater updater = blockchainStorage.updater();
+      long headerNumber = startBlockNumber;
+      final long endBlockNumber =
+          Math.min(startBlockNumber + preMergeBlockPrunerRange, mergeBlockNumber);
+      do {
+        blockchainStorage
+            .getBlockHash(headerNumber)
+            .filter((h) -> blockchainStorage.getBlockBody(h).isPresent())
+            .ifPresent(
+                (h) -> {
+                  updater.removeTransactionReceipts(h);
+                  updater.removeTotalDifficulty(h);
+                  blockchainStorage
+                      .getBlockBody(h)
+                      .map((bb) -> bb.getTransactions())
+                      .ifPresent(
+                          (transactions) ->
+                              transactions.stream()
+                                  .map(Transaction::getHash)
+                                  .forEach((th) -> updater.removeTransactionLocation(th)));
+                  updater.removeBlockBody(h);
+                });
+      } while (++headerNumber < endBlockNumber);
+      updater.commit();
+
+      if (endBlockNumber % preMergeBlockPrunerRange * 1000 == 0) {
+        LOG.info("Pruned up to block {}", endBlockNumber);
+      }
+
+      try {
+        Thread.sleep(preMergeBlockPrunerSleepTime);
+      } catch (InterruptedException e) {
+        return;
```

Comment:
Would this result in the Pruner thread dying? 
Think we should log an error here at least and/or possibility to recover?